internal/utils.go

```diff
@@ -32,6 +44,172 @@ func CreateFullOptions(max int) Options {
+			}
+			return nil, nil, err
+		}
+		currPid := pkt.PID()
+		if currPid != pid {
+			continue
+		}
+		packets = append(packets, *pkt)
+
+		_, err = pmtAcc.WritePacket(pkt)
+		if err == gots.ErrAccumulatorDone {
+			pmt, err = psi.NewPMT(pmtAcc.Bytes())
+			if err != nil {
+				return nil, nil, err
+			}
+			if len(pmt.Pids()) == 0 {
+				done = false
+				pmtAcc = packet.NewAccumulator(psi.PmtAccumulatorDoneFunc)
+				continue
+			}
+			done = true
+		} else if err != nil {
+			return nil, nil, err
+		}
+	}
+
+	return packets, pmt, nil
+}
+
+func WritePacket(pkt *packet.Packet, w io.Writer) error {
+	_, err := w.Write(pkt[:])
+	return err
+}
+
+func RemoveFileIfExists(file string) error {
+	// Remove the file if it exists
+	if _, err := os.Stat(file); errors.Is(err, os.ErrNotExist) {
+		// file does not exist
+		return nil
+	}
+
+	err := os.Remove(file)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func OpenFileAndAppend(file string) (*os.File, error) {
+	// Create and append to the new file
+	fo, err := os.OpenFile(file, os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0644)
+	if err != nil {
+		return nil, fmt.Errorf("creating output file %w", err)
+	}
+
+	return fo, nil
+}
+
+func ParseAstitsElementaryStreamInfo(es *astits.PMTElementaryStream) *ElementaryStreamInfo {
+	var streamInfo *ElementaryStreamInfo
+	switch es.StreamType {
+	case astits.StreamTypeH264Video:
+		streamInfo = &ElementaryStreamInfo{PID: es.ElementaryPID, Codec: "AVC", Type: "video"}
+	case astits.StreamTypeAACAudio:
+		streamInfo = &ElementaryStreamInfo{PID: es.ElementaryPID, Codec: "AAC", Type: "audio"}
+	case astits.StreamTypeH265Video:
+		streamInfo = &ElementaryStreamInfo{PID: es.ElementaryPID, Codec: "HEVC", Type: "video"}
+	case astits.StreamTypeSCTE35:
+		streamInfo = &ElementaryStreamInfo{PID: es.ElementaryPID, Codec: "SCTE35", Type: "cue"}
+	}
+
+	return streamInfo
+}
+
+func ParseElementaryStreamInfo(es psi.PmtElementaryStream) *ElementaryStreamInfo {
+	pid := uint16(es.ElementaryPid())
+	var streamInfo *ElementaryStreamInfo
+	switch es.StreamType() {
+	case psi.PmtStreamTypeMpeg4VideoH264:
+		streamInfo = &ElementaryStreamInfo{PID: pid, Codec: "AVC", Type: "video"}
+	case psi.PmtStreamTypeAac:
+		streamInfo = &ElementaryStreamInfo{PID: pid, Codec: "AAC", Type: "audio"}
+	case psi.PmtStreamTypeMpeg4VideoH265:
+		streamInfo = &ElementaryStreamInfo{PID: pid, Codec: "HEVC", Type: "video"}
+	case psi.PmtStreamTypeScte35:
+		streamInfo = &ElementaryStreamInfo{PID: pid, Codec: "SCTE35", Type: "cue"}
+	}
+
+	return streamInfo
+}
+
+func ParsePatcketToPAT(pkt *packet.Packet) (pat psi.PAT, e error) {
+	if packet.IsPat(pkt) {
+		pay, err := packet.Payload(pkt)
+		if err != nil {
+			return nil, err
+		}
+
+		pat, err = psi.NewPAT(pay)
+		if err != nil {
+			return nil, err
+		}
+
+		return pat, nil
+	}
+
+	return nil, fmt.Errorf("unable to parse packet to PAT")
+}
+
+// Check if two sets contain same elements
+func IsTwoSetsOverlapping(s1 []int, s2 []int) bool {
```

Comment:
The normal way to make a set in Go is to use a map[int]struct{}. The map makes it impossible to store the same value twice, and the empty struct is there since it takes no space to store so it's cheaper than a bool that takes a byte. I think your implementation here is fine for its purpose, though.